### ISSUE TITLE
ambient_authority feature

### ISSUE THREAD

**Issue by ratmice** [bytecodealliance/cap-std#344]
When using cap_std in libraries, it would be nice to be able to separate libraries using ambient authority from libraries which do not.
In order to maintain additive features, the feature would need to enable ambient authority.  Because the cat is out of the bag, I believe the appropriate way to do this would require disabling `default-features` to disable ambient authority. Then enable the `ambient_authority` feature in the libraries that need it.

```
default-features = ["ambient_authority"]
```

Should the ambient_authority crate have a similar mechanism that essentially turn it into a no-op crate?
I imagine that that feature might make implementing this in cap_std easier since it allows you to turn uses into errors.

I would be happy to work on a patch, if this sounds alright.

**Comment by sunfishcode** [comment]
I'm not completely clear on how this would work, but it seems to me that if you want the ambient-authority crate to be conditionally empty, a better approach would be to just depend on the ambient-authority crate using `optional = true`.

**Comment by ratmice** [comment]
Yeah that makes sense regarding the a-a crate.

I will make an exploratory repository w/ a binary/library (depending directly/optionally on ambient-authority) to show how it would work in cap_std, which will be much less involved than trying to explain it on something `cap_std` sized.

**Comment by ratmice** [comment]
FWIW, seems like this might not be possible to do without a version bump due to dependencies `default-features = false`, currently having `ambient_authority`.

**Comment by ratmice** [comment]
Anyhow in making my test it brought up a bunch of edge cases in the cargo feature resolver,
it works in some situations (like the ones i'm actually targeting) where I have a bunch of things which depend upon cap_std types,
but don't actually exercise any ambient authority themselves.

Then I have another thing which depends upon it that does depend upon ambient authority.
But once you throw these two things into a workspace, or have multiple dependencies resolving different features.
The thing which has the feature disabled, will either (if it references ambient-authority) compile or fail depending upon which directory in the workspace you run cargo from.

In other words it sometimes works, but it's totally unreliable in the sense that features seem to be enabled based upon the union of the workspaces dependency graph, not the dependency graph of each individual member.
So, I think it is better not to have this feature than to have it and be totally unreliable.

**Comment by ratmice** [comment]
I threw the example here, perhaps I could be just missing something dumb:

https://github.com/ratmice/ubiquitous-lamp

Essentially if you have non_ambient_consumer on it's own in it's own repository and if it doesn't depend upon anything which enables the "ambient-authority" feature of the `ambient_provider` lib, it will fail as it does when you `cd non_ambient_consumer` build, but you can still depend upon it from something which enables the `ambient-authority` feature.

Once you throw it in a workspace like this though, hilarity ensues.

**Comment by sunfishcode** [comment]
Yeah, it's my understanding that the uninoing is the reason why features are supposed to be additive. When you depend on a crate with one set of features, it may get more features enabled by other crates that also depend on it, so everything needs to be additive so that doesn't break.

If all your crates use the ambient-authority crate to at least indicate *where* they use ambient authorities, as cap-std does, then one alternative is to use [the clippy approach](https://github.com/sunfishcode/ambient-authority/blob/main/clippy.toml) to scan for uses of ambient authority.
